nx/lib/nx/defn/evaluator.ex

```diff
@@ -469,4 +477,51 @@ defmodule Nx.Defn.Evaluator do
   defp composite_to_params(other, acc) do
     [fn -> other end | acc]
   end
+
+  # Debug hook: print/save node info if enabled and not already printed
+  defp maybe_debug_node(ans, res, %{debug_options: nil} = state), do: {res, state}
+  defp maybe_debug_node(%Nx.Tensor{data: %Expr{id: id, op: op, args: args}} = ans, res, %{debug_options: opts, printed_nodes: printed_nodes} = state) do
```

Comment:
Let's move the expr destructuring inside the function because it doesn't affect control flow. We can keep it as data: expr here